sql/analyzer/hoist_select_exists.go

```diff
@@ -40,140 +40,130 @@ func hoistSelectExists(
+	ret := filter.Child
+	var retFilters []sql.Expression
+	same := transform.SameTree
+	for _, f := range splitConjunction(filter.Expression) {
+		var joinType plan.JoinType
+		var s *hoistSubquery
 		switch e := f.(type) {
 		case *plan.ExistsSubquery:
-			decorrelated, outerFilters = decorrelateOuterCols(e.Query, scopeLen)
-			if len(outerFilters) == 0 {
-				return nil
+			s = decorrelateOuterCols(e.Query, scopeLen)
+			if s == nil {
+				return filter, transform.SameTree, nil
 			}
 			joinType = plan.JoinTypeSemi
 		case *expression.Not:
 			esq, ok := e.Child.(*plan.ExistsSubquery)
 			if !ok {
-				return nil
+				return filter, transform.SameTree, nil
 			}
-			decorrelated, outerFilters = decorrelateOuterCols(esq.Query, scopeLen)
-			if len(outerFilters) == 0 {
-				return nil
+			s = decorrelateOuterCols(esq.Query, scopeLen)
+			if s == nil {
+				return filter, transform.SameTree, nil
 			}
 			joinType = plan.JoinTypeAnti
 		default:
+			retFilters = append(retFilters, f)
+			continue
 		}
-	}
-	if len(outerFilters) == 0 {
-		return nil
-	}
-	if len(newFilters) == 0 {
-		return &hoistExistsSubquery{
-			left:         filter.Child,
-			right:        decorrelated,
-			outerFilters: outerFilters,
-			joinType:     joinType,
+
+		// if we reached here, |s| contains the state we need to
+		// decorrelate the subquery expression into a new node
+		outerFilters, _, err := FixFieldIndexesOnExpressions(scope, a, append(ret.Schema(), s.inner.Schema()...), s.outerFilters...)
+		if err != nil {
+			return filter, transform.SameTree, err
 		}
+
+		retFilters = append(retFilters, s.innerFilters...)
+
+		switch joinType {
+		case plan.JoinTypeAnti:
+			ret = plan.NewAntiJoin(ret, s.inner, expression.JoinAnd(outerFilters...))
+		case plan.JoinTypeSemi:
+			ret = plan.NewSemiJoin(ret, s.inner, expression.JoinAnd(outerFilters...))
+		default:
+			panic("expected JoinTypeSemi or JoinTypeAnti")
+		}
+		same = transform.NewTree
 	}
-	newFilter := plan.NewFilter(expression.JoinAnd(newFilters...), filter.Child)
 
-	return &hoistExistsSubquery{
-		left:         newFilter,
-		right:        decorrelated,
-		outerFilters: outerFilters,
-		joinType:     joinType,
+	if same {
+		return filter, transform.SameTree, nil
+	}
+	if len(retFilters) > 0 {
+		ret = plan.NewFilter(expression.JoinAnd(retFilters...), ret)
 	}
+	return ret, transform.NewTree, nil
+}
+
+type hoistSubquery struct {
+	inner        sql.Node
+	innerFilters []sql.Expression
+	outerFilters []sql.Expression
 }
 
 // decorrelateOuterCols returns an optionally modified subquery and extracted
 // filters referencing an outer scope.
-func decorrelateOuterCols(e *plan.Subquery, scopeLen int) (sql.Node, []sql.Expression) {
+func decorrelateOuterCols(e *plan.Subquery, scopeLen int) *hoistSubquery {
 	var outerFilters []sql.Expression
+	var innerFilters []sql.Expression
 	n, same, _ := transform.Node(e.Query, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
```

Comment:
`transform.Node` here is bottom up and this seems to work only if there is exactly one `*Filter` node that we visit and it applies to basically the whole subtree. That might be true given our current opaque nodes but I just want to be sure...There's no way for us to get a Filter() node in the middle of a join tree where we're evaluating this, right? (I think I saw some comments about something like our handling of Natural joins or something doing that currently...) And we don't traverse into subquery aliases that might appear as table factors in a join? Or into unions where one half has a filter?

Reply:
I rearranged the rule order to keep join trees clear of projections and filters, including through subquery/[expr] resolve. Natural joins introduce projections into join trees, which is also problem. Both join planning and `transform.Node` stop at opaque boundaries, but not maybe for the same reason. Join planning expects a join tree to have `*plan.JoinNode` or `sql.NameableNode`, discarding filler nodes. Basically "yes" for all of your questions, though it is possible our tests miss edge cases we haven't thought of.